### ISSUE TITLE
Type improperly handled when variable passed by reference to lambda.

### ISSUE THREAD

**Issue by rzvc** [vimeo/psalm#7939]
This issue has been raised multiple times, and rejected each time. I believe the idea that it's impossible to be handled properly is wrong, which is why I'm opening this new ticket, where I'm trying to break it down into one enhancement and two bugs, which when implemented/fixed should solve this problem.

The code:
```php
<?php
/** @var string[] */
$arr = [];

// Suggested behaviour:
//	- Type of $arr inside the lambda: string[] | 123 | true.
//	- Type of $arr doesn't change when this is called.
$callback1 = function() use (&$arr) : void
{
};

// <-- Bug 1:
// Currently type of $arr at this point is mixed. Since $callback1 is not even being called,
// the type should continue to be string[].

$arr = true;
$callback1();

// <-- Bug 2:
// Since $callback1() is not changing $arr in any way, $arr's type should remain untouched.
// Instead, currently the type gets thrashed again, and it becomes mixed at this point.

// Suggested behaviour:
//	- Type of $arr inside the lambda: string[] | 123 | true.
//	- Type of $arr becomes int after a call to this function.
$callback2 = function() use (&$arr) : void
{
	$arr = 123;
};

// <-- Bug 1 also present here.

$arr = true;
$callback2();

// <-- Bug 2 - the type of arr becomes mixed again, when in fact it should be 123.

// Suggested behaviour:
//	- Type of $arr inside the lambda: string[] | 123 | true.
//	- Type of $arr becomes T | int, where T is whatever type $arr is at the time $callback3 is called.
$callback3 = function() use (&$arr) : void
{
    if (rand())
		$arr = 123;
};

// <-- Bug 1.

$arr = true;
$callback3();

// <-- Bug 2. $arr should be true | 123 at this point.
```
https://psalm.dev/r/4a2996c9fb

The enhancement is that inside the lambda, $arr would end up having a type that would be the union of all the visible types of $arr. This might not be necessary at all times, but it would simplify things a lot if we considered that it is. If it's not obvious, the reason it is necessary is because the callback could leave scope and be called by other code at unknown points, as we're advancing through the current scope.

Bug 1 addresses the fact that the type gets thrashed by the mere mention of a lambda - this should be the easiest one to fix.

Bug 2 replaces the current behaviour, of thrashing the type of the referenced variable, with augmenting/replacing it, based on the statements acting on that variable inside the lambda. This might require re-analyzing the lambda each time it is referenced.

Previous issues that I found on this topic: #2391 #3856 #3688 #4536

Edit:
I have been thinking a little more about it and the callback leaving the scope is more of an issue than I initially thought. If that can be detected in some way, then the type of the referenced variable at any later time needs to be &lt;last assigned type&gt; | &lt;type resulted from calling the callback on the last assigned type&gt;.

If it cannot be detected, then &lt;last assigned type&gt; | &lt;type resulted from calling the callback on the last assigned type&gt; everywhere is better than mixed.

**Comment by rzvc** [comment]
If it's going to be analyzed only once, then ```array|bool|string``` would be great - definitely much better than ```mixed```. Afterwards, whenever the closure gets referenced, the type of the variables can be augmented/changed with the types that resulted from the initial analysis of the closure (in the example you showed, it would be ```(T) => bool```, while in my $callback3 example, it would become (T) => T | 123).

I'm also open to some type of @strict-var, like you suggested, since it would be a huge improvement over the current situation, where you have to redeclare the variable inside the closure, after the closure declaration, and after each closure reference (this indicates that references are being tracked, and augmentation is possible, solving Bug 2).

**Comment by danog** [comment]
Implemented!
